### PR TITLE
Introduce pre-commit for MQT DDSIM

### PR DESCRIPTION
**Pull request by hillmich**
This PR introduces pre-commit to MQT DDSIM with a selected set of checks to begin with.

Many thanks @burgholzer !

### CODE REVIEW

**Comment by hillmich** [comment]
@burgholzer Could you review please?

File: docs/source/conf.py [hunk 1]
```diff
@@ -1,14 +1,12 @@
 from importlib.metadata import version
 
-
 # -- Project information -----------------------------------------------------
-project = 'DDSIM'
-copyright = '2022, Stefan Hillmich'
```
Comment (burgholzer) [inline docs/source/conf.py:6]:
```python
project_copyright = "Chair for Design Automation, Technical University of Munich"
```

File: docs/source/simulators/SimulationPathFramework.rst [hunk 2]
```diff
@@ -49,9 +49,9 @@ Per default, the backend uses the :code:`sequential` strategy. For this particul
  - then, the first CNOT is applied (:code:`[4, 2] -> 5`)
  - finally, the last CNOT is applied (:code:`[5, 3] -> 6`)
 
- The corresponding simulation path is thus described by :code:`[[0, 1], [4, 2], [5, 3]]` and the final state is the one with ID :code:`6`.
+ The corresponding simulation path is thus described by :code:``[[0, 1], [4, 2], [5, 3]]`` and the final state is the one with ID :code:`6`.
```
Comment (burgholzer) [inline docs/source/simulators/SimulationPathFramework.rst:52]: Not quite sure the double backticks are intentional here. RST tags work with single backticks.

File: mqt/ddsim/CMakeLists.txt [hunk 3]
```diff
@@ -1,3 +1,5 @@
 pybind11_add_module(py${PROJECT_NAME} bindings.cpp)
-target_link_libraries(py${PROJECT_NAME} PUBLIC ${PROJECT_NAME} MQT::qfr_python pybind11_json)
-target_compile_definitions(py${PROJECT_NAME} PRIVATE VERSION_INFO=${DDSIM_VERSION_INFO})
+target_link_libraries(py${PROJECT_NAME} PUBLIC ${PROJECT_NAME} MQT::qfr_python
+                                               pybind11_json)
```
Comment (burgholzer) [inline mqt/ddsim/CMakeLists.txt:3]:
```suggestion
target_link_libraries(py${PROJECT_NAME} PUBLIC ${PROJECT_NAME} MQT::qfr_python)
```
pybind11_json is automatically linked with the `qfr_python` target.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -1,46 +1,57 @@
...
+  ${PROJECT_SOURCE_DIR}/include/PathSimulator.hpp
+  ${CMAKE_CURRENT_SOURCE_DIR}/PathSimulator.cpp)
+target_include_directories(
+  ${PROJECT_NAME}
+  PUBLIC $<BUILD_INTERFACE:${${PROJECT_NAME}_SOURCE_DIR}/include>)
 # set required C++ standard and disable compiler specific extensions
 target_compile_features(${PROJECT_NAME} PUBLIC cxx_std_17)
-set_target_properties(${PROJECT_NAME} PROPERTIES CMAKE_CXX_STANDARD_REQUIRED ON CXX_EXTENSIONS OFF)
+set_target_properties(${PROJECT_NAME} PROPERTIES CMAKE_CXX_STANDARD_REQUIRED ON
+                                                 CXX_EXTENSIONS OFF)
```
Comment (burgholzer) [inline src/CMakeLists.txt:32]: I am fairly sure that these lines are no longer required as they are covered by the shared project setting from the DDPackage

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -1,71 +1,96 @@
...
+    gtest_force_shared_crt
+    gtest_hide_internal_symbols)
+  # cmake-lint: disable=C0307
+  set_target_properties(gtest gtest_main gmock gmock_main PROPERTIES FOLDER
+                                                                     extern)
+  if(BINDINGS)
+    # adjust visibility settings for building Python bindings
+    target_compile_options(gtest PUBLIC -fvisibility=hidden)
+    target_compile_options(gmock PUBLIC -fvisibility=hidden)
+  endif()
```
Comment (burgholzer) [inline test/CMakeLists.txt:27]: these lines can all be removed since the latest DDPackage update

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -1,71 +1,96 @@
...
+      CACHE BOOL "Skip google benchmark self tests")
+  add_subdirectory(
+    "${PROJECT_SOURCE_DIR}/extern/qfr/extern/dd_package/extern/benchmark"
+    "extern/qfr/extern/dd_package/extern/benchmark" EXCLUDE_FROM_ALL)
+  set_target_properties(benchmark benchmark_main
+                        PROPERTIES FOLDER extern/qfr/extern/dd_package/extern)
+  if(BINDINGS)
+    # adjust visibility settings for building Python bindings
+    target_compile_options(benchmark PUBLIC -fvisibility=hidden)
+  endif()
```
Comment (burgholzer) [inline test/CMakeLists.txt:74]: can be removed since `pybind` no longer complains since recent versions

File: test/CMakeLists.txt [hunk 7]
```diff
@@ -1,71 +1,96 @@
...
+  set(THREADS_PREFER_PTHREAD_FLAG ON)
+  find_package(Threads REQUIRED)
+  set(BENCHMARK_ENABLE_TESTING
+      OFF
+      CACHE BOOL "Skip google benchmark self tests")
+  add_subdirectory(
+    "${PROJECT_SOURCE_DIR}/extern/qfr/extern/dd_package/extern/benchmark"
+    "extern/qfr/extern/dd_package/extern/benchmark" EXCLUDE_FROM_ALL)
+  set_target_properties(benchmark benchmark_main
+                        PROPERTIES FOLDER extern/qfr/extern/dd_package/extern)
```
Comment (burgholzer) [inline test/CMakeLists.txt:70]: can also be removed as the file is automatically put there.

File: test/CMakeLists.txt [hunk 8]
```diff
@@ -1,71 +1,96 @@
...
+               ${CMAKE_CURRENT_SOURCE_DIR}/benchmark.cpp)
+target_link_libraries(ddsim_benchmark PRIVATE ${PROJECT_NAME}
+                                              benchmark::benchmark_main)
 set_target_properties(ddsim_benchmark PROPERTIES EXPORT_NAME ddsim_benchmark)
-set_target_properties(ddsim_benchmark PROPERTIES FOLDER tests CMAKE_CXX_STANDARD_REQUIRED ON CXX_EXTENSIONS OFF)
+set_target_properties(
+  ddsim_benchmark
+  PROPERTIES FOLDER tests
+             CMAKE_CXX_STANDARD_REQUIRED ON
+             CXX_EXTENSIONS OFF)
```
Comment (burgholzer) [inline test/CMakeLists.txt:86]: should not be necessary any more

File: test/python/hybridsimulator/test_hybrid_qasm_simulator.py [hunk 9]
```diff
@@ -21,60 +20,58 @@ def setUp(self):
...
         """Test backend.status()."""
-        status = self.backend.status()
-        return status
+        return self.backend.status()
 
     def test_qasm_simulator_single_shot(self):
         """Test single shot run."""
         result = execute(self.circuit, self.backend, shots=1).result()
-        self.assertEqual(result.success, True)
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/hybridsimulator/test_hybrid_qasm_simulator.py:42]:
```suggestion
        assert result.success
```

File: test/python/hybridsimulator/test_hybrid_qasm_simulator.py [hunk 10]
```diff
@@ -86,8 +83,8 @@ def test_dd_mode_simulation(self):
         circ.measure_all(inplace=True)
         print(circ.draw(fold=-1))
         self.circuit = circ
-        result = execute(self.circuit, self.backend, mode='dd').result()
-        self.assertEqual(result.success, True)
+        result = execute(self.circuit, self.backend, mode="dd").result()
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/hybridsimulator/test_hybrid_qasm_simulator.py:87]:
```suggestion
        assert result.success
```

File: test/python/hybridsimulator/test_hybrid_qasm_simulator.py [hunk 11]
```diff
@@ -100,6 +97,6 @@ def test_amplitude_mode_simulation(self):
         circ.measure_all(inplace=True)
         print(circ.draw(fold=-1))
         self.circuit = circ
-        result = execute(self.circuit, self.backend, mode='amplitude').result()
-        self.assertEqual(result.success, True)
+        result = execute(self.circuit, self.backend, mode="amplitude").result()
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/hybridsimulator/test_hybrid_qasm_simulator.py:101]:
```suggestion
        assert result.success
```

File: test/python/hybridsimulator/test_hybrid_statevector_simulator.py [hunk 12]
```diff
@@ -1,46 +1,44 @@
...
         """Test backend.status()."""
-        status = self.backend.status()
-        return status
+        return self.backend.status()
 
     def test_statevector_output(self):
         """Test final state vector for single circuit run."""
         result = execute(self.q_circuit, backend=self.backend, shots=0).result()
-        self.assertEqual(result.success, True)
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/hybridsimulator/test_hybrid_statevector_simulator.py:35]:
```suggestion
        assert result.success
```

File: test/python/simulator/test_qasm_simulator.py [hunk 13]
```diff
@@ -21,57 +20,55 @@ def setUp(self):
...
         """Test backend.status()."""
-        status = self.backend.status()
-        return status
+        return self.backend.status()
 
     def test_qasm_simulator_single_shot(self):
         """Test single shot run."""
         result = execute(self.circuit, self.backend, shots=1).result()
-        self.assertEqual(result.success, True)
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/simulator/test_qasm_simulator.py:42]:
```suggestion
        assert result.success
```

File: test/python/simulator/test_statevector_simulator.py [hunk 14]
```diff
@@ -1,42 +1,40 @@
...
         """Test backend.status()."""
-        status = self.backend.status()
-        return status
+        return self.backend.status()
 
     def test_statevector_output(self):
         """Test final state vector for single circuit run."""
         result = execute(self.q_circuit, backend=self.backend).result()
-        self.assertEqual(result.success, True)
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/simulator/test_statevector_simulator.py:33]:
```suggestion
        assert result.success
```

File: test/python/taskbasedsimulator/test_path_sim_qasm_simulator.py [hunk 15]
```diff
@@ -20,93 +20,91 @@ def setUp(self):
...
         """Test backend.status()."""
-        status = self.backend.status()
-        return status
+        return self.backend.status()
 
     def test_qasm_simulator_single_shot(self):
         """Test single shot run."""
         result = execute(self.circuit, self.backend, shots=1).result()
-        self.assertEqual(result.success, True)
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/taskbasedsimulator/test_path_sim_qasm_simulator.py:42]:
```suggestion
        assert result.success
```

File: test/python/taskbasedsimulator/test_path_sim_statevector_simulator.py [hunk 16]
```diff
@@ -1,54 +1,53 @@
...
         """Test backend.status()."""
-        status = self.backend.status()
-        return status
+        return self.backend.status()
 
     def test_statevector_output(self):
         """Test final state vector for single circuit run."""
         result = execute(self.q_circuit, backend=self.backend).result()
-        self.assertEqual(result.success, True)
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/taskbasedsimulator/test_path_sim_statevector_simulator.py:33]:
```suggestion
        assert result.success
```

File: test/python/taskbasedsimulator/test_path_sim_statevector_simulator.py [hunk 17]
```diff
@@ -1,54 +1,53 @@
...
+        assert actual[1] == 0
+        assert actual[2] == 0
+        assert math.isclose((abs(actual[3])) ** 2, 0.5, abs_tol=0.0001)
 
     def test_statevector_output_pairwise(self):
         """Test final state vector for single circuit run."""
         mode = "pairwise_recursive"
         result = execute(self.q_circuit, backend=self.backend, mode=mode).result()
-        self.assertEqual(result.success, True)
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/taskbasedsimulator/test_path_sim_statevector_simulator.py:46]:
```suggestion
        assert result.success
```

File: test/python/unitarysimulator/test_unitary_simulator.py [hunk 18]
```diff
@@ -1,48 +1,46 @@
...
+        assert np.count_nonzero(result.get_unitary("test")) == self.non_zeros_in_bell_circuit
         return result
 
     def test_unitary_simulator_recursive_mode(self):
-        result = execute(self.circuit, self.backend, mode='recursive').result()
-        self.assertEqual(result.success, True)
-        print(result.data('test'))
-        self.assertEqual(np.count_nonzero(result.get_unitary('test')), 16)
+        result = execute(self.circuit, self.backend, mode="recursive").result()
+        assert result.success is True
```
Comment (burgholzer) [inline test/python/unitarysimulator/test_unitary_simulator.py:43]:
```suggestion
        assert result.success
```

**Review by burgholzer — CHANGES_REQUESTED** [review]
Many thanks for your work. Just went through all the changes and found a few minor things that should be easy to fix.
I'll quickly batch up all the quick-fix suggestions and commit them right away so it's less effort for you.
The addition of the other checks should merely create some whitespace changes and potentially fix some spelling errors.

**Comment by hillmich** [comment]
Took a few minutes to get black and ruff to agree on a function with "too many" arguments :sweat_smile:

File: pyproject.toml [hunk 19]
```diff
@@ -93,3 +93,61 @@ environment = { MACOSX_DEPLOYMENT_TARGET = "10.15", DEPLOY = "ON" }
...
+    "PTH",         # flake8-use-pathlib
+    "Q",           # flake8-quotes
+    "RET",         # flake8-return
+    "RUF",         # Ruff-specific
+    "SIM",         # flake8-simplify
+    "TCH",         # flake8-type-checking
+    "UP",          # pyupgrade
+    "YTT",         # flake8-2020
+]
+ignore = ["E501"]
```
Comment (burgholzer) [inline pyproject.toml:123]: We could just add `"PLR2004", "PLR0913",` here as well. (magic numbers and too many arguments)
Just looked into other repositories and these two rules seem to be ignored in most of them.
```suggestion
ignore = [
  "PLR2004", # Magic values
  "PLR0913", # Too many arguments
  "E501",    # Line too long (Black is enough)
]
```

**Review by burgholzer — CHANGES_REQUESTED** [review]
Just a couple of small little details.
I think you should be able to just batch these suggestions up on GitHub and apply them (if you agree with them of course 🙃)
The config change might require running pre-commit once though (as it should eliminate some `noqa` statements.
Then, this is good to go.

**Review by burgholzer — APPROVED** [review]

**Pull request merged**